src/test/java/redis/clients/jedis/tests/ShardedJedisPoolTest.java

```diff
@@ -103,6 +105,53 @@ public void checkPoolRepairedWhenJedisIsBroken() {
     pool.destroy();
   }
 
+  @Test
+  public void checkRepairedWhenJedisIsBroken() {
+    ShardedJedisPool pool = new ShardedJedisPool(new GenericObjectPoolConfig(), shards);
+    JedisShardInfo shardInfo0 = shards.get(0);
+    JedisShardInfo shardInfo1 = shards.get(1);
+    ShardedJedis jedis1 = pool.getResource();
+    Jedis shard10 = jedis1.getShard(shardInfo0);
+    Jedis shard11 = jedis1.getShard(shardInfo1);
+    breakClient(shard10);
+    assertTrue(shard10.getClient().isBroken());
+    jedis1.close();
+
+    ShardedJedis jedis2 = pool.getResource();
+    assertEquals(jedis1, jedis2);
+    Jedis shard20 = jedis2.getShard(shardInfo0);
+    Jedis shard21 = jedis2.getShard(shardInfo1);
+    assertFalse(shard20.getClient().isBroken());
+    assertNotSame(shard10, shard20);
+    assertSame(shard11, shard21);
+    jedis2.close();
+    pool.destroy();
+  }
+
+  @Test
+  public void checkRepairedWhenJedisIsBroken2() {
+    GenericObjectPoolConfig config = new GenericObjectPoolConfig();
+    config.setTestOnBorrow(true);
+    List<JedisShardInfo> withBrokenShards = new ArrayList<JedisShardInfo>(shards);
+    withBrokenShards.set(0, new JedisShardInfo("localhost", 1234));
+    ShardedJedisPool pool = new ShardedJedisPool(config, withBrokenShards);
+    JedisShardInfo shardInfo0 = withBrokenShards.get(0);
+    JedisShardInfo shardInfo1 = withBrokenShards.get(1);
+    ShardedJedis jedis1 = pool.getResource();
+    Jedis shard10 = jedis1.getShard(shardInfo0);
+    Jedis shard11 = jedis1.getShard(shardInfo1);
+    jedis1.close();
+
+    ShardedJedis jedis2 = pool.getResource();
+    assertEquals(jedis1, jedis2);
```

Comment:
Same here: I'm not sure this is guaranteed by Apache Commons Pool. Maybe need to change the pool config to have size 1 to ensure this.